spec/message.abnf

```diff
@@ -81,7 +86,7 @@ name-start = ALPHA / "_"
 name-char  = name-start / DIGIT / "-" / "." / ":"
            / %xB7 / %x300-36F / %x203F-2040
 
-text-escape     = backslash ( backslash / "{" / "}" )
+text-escape     = backslash ( backslash / "." / "{" / "}" )
```

Comment:
> Do we need `\.` all the time? Or can we make it a special case of starting a simple message?

Let's _please_ avoid such special cases; they're unnecessary complexity for people and machines alike. If `\.` is a valid simple message representing ".", then `period\.` and `\.\.\.` should be valid as well (respectively representing "period." and "...").

> I'm almost willing to suggest that we require patterns starting with `.` to be quoted and not provide any escape

I had the same thought, and that approach I could live with.

Reply:
> I'm almost willing to suggest that we require patterns starting with . to be quoted and not provide any escape

+1. We need to support `{{.com}}` anyways as a valid pattern, so we could reduce the scope of this PR, and not introduce any new escape sequences. Then disccuss separately.

Reply:
Not supporting `\.` feels to me like it would introduce an odd disparity in how the two code-y characters `{` and `.` are handled when they show up in a pattern. With `{`, the answer is to escape it as `\{`. But with `.` I need to escape it as `{{.}}`? Why not support `\.` as well?

Reply:
OTOH, one could also argue that *supporting* `\.` introduces an odd disparity in how the period in column 0 needs to be quoted but any other period doesn't. Or that it's an odd diparity in how `\.` produces a `.` in column 0 but it isn't allowed anywhere else (depending on the choice we make in this thread).

This is a good illustration of the accidental comprexity brought about by choosing D in #525. Note that neither A, E, nor F suffer from this problem.

Reply:
That's not correct, though. Every option has column 0 quoting issues. They just differ in frequency and such. Options C and E were about trying to minimize the frequency.

Reply:
In the options using `{` as the first character of the code mode introducer sequence, the quoting issues are not specific to column 0, and do not create inconsistencies in columns 1 and onwards. They leverage the exact same escaping mechanism: `\{`.

I know this may seem like a banal observation, but I don't think the consequences of having separate escaping requirements were thoroughly discussed. What's your take on what should happen to `foo\.`?

Reply:
No, @stasm is right. A, E, and F all used prefixes that start with `{`, which already needs escaping _everywhere_ in pattern text. So,
* In option A, a simple message producing `{{Use pattern quotes for messages}}` would look like e.g. `\{\{Use pattern quotes for messages\}\}`.
* In option E, a simple message producing `{^} Use a magic sequence at the start` would look like e.g. `\{^\} Use a magic sequence at the start`.
* In option F, a simple message producing `{% Use a preamble with %} {{Patterns to follow}}` would look like e.g. `\{% Use a preamble with %\} \{\{Patterns to follow\}\}`.

And in none of those is any special treatment necessary to escape the first character of a simple message that is not also necessary everywhere else the same character appears in it.

Reply:
Given that we _have_ selected option D, could we focus in this thread on whether to allow for `\.` to represent a literal `.` in a pattern, rather than rehashing what the situation would have been had we chosen something else?

My position is that given that we already allow for escaping `{` as `\{` and `}` as `\}`, some users will have an expectation that they may also escape `.` as `\.`, and that with `.` as a keyword introducer, there is no sensible other meaning for `\.`. Therefore it behooves us to allow for it.

And if we don't allow for it, then by consistency we probably should also not allow for escaping `}` in a simple pattern, given that doing so is not required.

Reply:
Yes, but that's still an escape of `{` (the fact that we've accounted for it already doesn't make it "not an escape").

Anyway, one of the problems here is the sigil choice. Using a rarer sigil than `.` would help.

Some solutions might include:

- Reserve a bunch of keywords now. The problem with the current situation isn't the keywords we have, but rather future compatibility needs with keywords that we don't know. We could reserve a list of words for future standardization. (This solution is not a serious contender)

- Fix the sigils or "spelling". The problem here is that `.` is not rare and `..` is also not rare. If the starter sigil/sigil sequence is rare enough, while it still has to be escaped when in plain text, it isn't a serious problem for anyone. This could be done for existing keywords or only for future keywords.

Examples:

```
.=input {$var :foo}
.=local $bar = {$foo :function}
.?match {$var} {$bar}
[*] {{.= and .? are rare so users rarely have to escape them}}
[*] {{== and ?? or some other sigil choice instead of . suggests itself}}
[*] {{note that we started with #}}

.input {$var}
.local $bar = {$foo :function}
.=future 
{{future keywords start with a special sequence--sorry future me--
  and the well-known sequences .input/.local/.match are reserved now}}
```

- Enclosing instead of doubling. Make the keyword distinct in another way:

```
.input. {$var}
.local. $bar = {$foo :function}
~input~
!local!
```

- Adjust the syntax so there is no optional whitespace and the bracket is part of the word

```
.input{$var :function}
.local{$bar} = {$foo :function}
.match{$var}
[*] {{This is not a serious contender either}}
```

- Use the character we've already abused

```
{input} {$var :foo}
{local} $bar = {$foo :function}
{match} {$var} {$bar}
[*] {{So many \{ }}

Reply:
@aphillips I've hidden your preceding comment to keep this thread somewhat on-topic on whether and where to allow for `\` escaping the keyword introducer `.`.

Reply:
I think that we should not allow any `.` to be escaped except in the initial position. My previous comment was about making the escaping less necessary by making a clearer keyword syntax, since I think we want to allow future keyword additions. This design depends on code-mode-introducing keywords being something the parser can pick out. Modifying keywords to include a second sigil might help. 

I'm super nervous about escapes in general because of the interaction with host syntax escaping. In many languages `\.` becomes `\\.` in storage (this is the same conversation about quotes). The double `\\` can be consumed by tools and turned into plain text that is further escaped (`\\\\`). I'm not as worried about `\{` and `\|` because these will be rather rare (and `\|` only inside quoted literals which are mainly inside placeholders). But `\.` is our apostrophe nightmare waiting to happen?

Reply:
As discussed on today's call, the consensus resolution was to not include any `\.` escapes, so a message starting with a literal `.` needs to be quoted.